### DISCUSSION TITLE
Breaking Changes & Deprecation

### DISCUSSION THREAD

**Discussion by ful1e5** [projekt0n/github-nvim-theme#198] [Announcements]
Subscribe to this discussion to receive updates on breaking changes and deprecations.

**Comment by ful1e5** [comment]
88d43731d8cdded23fbf7dbdf314ae3934b84d82 includes the removal of all supportive modules from the **Vim configuration** that were using `vim.g.github_*`. Please migrate to the `setup` Lua function.

For further information about the deprecated modules and their replacements, please refer to `:h github-theme.deprecated` or [gt_deprecated.txt](https://github.com/projekt0n/github-nvim-theme/blob/main/doc/gt_deprecated.txt)

**Comment by ful1e5** [comment]
**Date: April 11, 2023**
**Ref:** 33369e97471dd716281a44030431da94f163f677

The `theme_style` option has been removed from the `setup` and the method of setting the colorscheme has changed. To set the colorscheme, use the `:colorscheme` command after calling setup.

Old approach for setting the theme:
```lua
  require('github-theme').setup({
    theme_style = 'dark',
    --- ...
  })
```

A new approach for setting the theme:
```lua
  require('github-theme').setup({
    --- ...
  })

  vim.cmd('colorscheme github_dark')
  -- or
  -- vim.cmd.colorscheme('github_dark')
```

**Comment by ful1e5** [comment]
**Date: April 12, 2023**
**Ref:** a5f6d561e3dbb84cb3c43731d9d433f62b9e5dc5

The configuration for defining **styles** has been migrated to `options.styles`. As part of this migration, `msg_area_style` has been removed.
```diff
require('github-theme').setup({
   ...
-  comment_style = "Italic",
-  function_style = "NONE",
-  keyword_style = "Italic",
-  variable_style = "NONE",
-  msg_are_style = "NONE",    -- Removed
+  options = {
+    styles = {
+      comments = "Italic",
+      functions = "NONE",
+      keywords = "Italic",
+      variables = "NONE"
+    },
+  },
   ...
}
```
Here are the updated config keys:
- `comment_style` -> `options.styles.comments`
- `function_style` -> `options.styles.functions`
- `keyword_style` -> `options.styles.keywords`
- `variable_style` -> `options.styles.variables`
- `msg_area_style` -> Removed.

**Comment by ful1e5** [comment]
**Date: April 23, 2023**
**Ref:** 1a96bd061cc30ab2543461d279baf87a68857ffe

The way to override color variables has been updated. You can now use the `specs` table to change colors for a specific colorscheme or for all of them with the `all` table. To override highlight groups, use the `groups` table in the configuration.

Some changes have been made to the names of the colorschemes. You can check the names of the colorschemes in the "colors" directory in the project root, or by using the `:colorscheme` vim command.

The deprecated colorschemes will no longer support any overriding highlight groups or color variables. The updated colorscheme list:

- github_dark_default -> github_dark_high_contrast
- github_dimmed -> github_dark_dimmed
- github_light_default -> github_light_high_contrast
- github_dark_tritanopia -> New!
- github_light_tritanopia -> New!

The updated configuration keys are:

- colors -> Use `specs` and `palettes` instead.
- overrides -> Use `groups` instead.

Example :

```lua
require('github-theme').setup({
  --- ...
  palettes = {
    github_dark = {
      canvas = {
        default = '#000000',
      },
    },
    all = {
      danger = {
        fg = '#FF0000',
      },
    },
  },
  specs = {
    github_dark_dimmed = {
      bg0 = '#111111',
    },
    all = {
      diag = {
        error = '#FF0000',
      },
    },
  },
  groups = {
    github_dark = {
      Error = { fg = '#FF0000' },
    },
    all = {
      ErrorMsg = { link = 'Error' },
    },
  },
})

vim.cmd.colorscheme('github_dark_colorblind')
```

**Comment by johannesrld** [comment]
Is there any particular reason to deprecate the dark default theme? dark_high_contrast is very clearly a different theme. It's quite annoying to have to manually remake dark_default because of this.

**Comment by tmillr** [comment]
**Date: July 23, 2024**
**Ref:** bc0c3953c71a8788b9d0511c7ef3410ab53917d5 (v1.1.0)

## Configuration & API

Highlight-group overrides/configuration have received a few small improvements. Assigning `false` or an empty table to a highlight group clears it. Assigning `false` to `groups`, `specs`, or `palettes` clears previous settings from the user config store for that category ("previous" meaning earlier in the same `nvim` session).

Setting `options.transparent` to `true` make's neovim's background transparent. Floats and sidebars will receive a similar background in this case, regardless of any other float/sidebar-specific options set (i.e. `options.transparent` overrides `options.darken.*`). See `:help github-nvim-theme-configuration`.

In the `Color` lib (`:help github-nvim-theme-color`), the `Color:lumanance()` method has been renamed to `Color:luminance()` (corrected typo/misspelling).

`options.darken.sidebars.enabled` has been renamed/reverted back to `options.darken.sidebars.enable` (see #343).

## Highlight Groups

As of neovim `v0.10`, several highlight groups have changed. You may need to update your config if you are using or overriding any of the groups listed in the following sections. You can also view the affected groups directly in neovim with `:help github-theme.nvim-0.10.0.hlgroups`.

#### Highlight Groups Renamed

<details>
<summary>Show groups</summary>
<br />

| Old Group                          | New Group                                                                             |
| ---------------------------------- | ------------------------------------------------------------------------------------- |
| `@conditional`                     | `@keyword.conditional`                                                                |
| `@debug`                           | `@keyword.debug`                                                                      |
| `@define`                          | `@keyword.directive` or `@keyword.directive.define`                                                      |
| `@exception`                       | `@keyword.exception`                                                                  |
| `@field`                           | `@variable.member`                                                                    |
| `@float`                           | `@number.float`                                                                       |
| `@include`                         | `@keyword.import`                                                                     |
| `@method`                          | `@function.method`                                                                    |
| `@method.call`                     | `@function.method.call`                                                               |
| `@namespace`                       | `@module`                                                                             |
| `@parameter`                       | `@variable.parameter`                                                                 |
| `@preproc`                         | `@keyword.directive`                                                                  |
| `@punctuation.special`             | `@markup.list` (markdown lists only; move subitems from `@text.todo`)                 |
| `@repeat`                          | `@keyword.repeat`                                                                     |
| `@storageclass`                    | `@keyword.storage`                                                                    |
| `@string.regex`                    | `@string.regexp`                                                                      |
| `@string.special`                  | `@markup.link.label` (non-url links)                                                  |
| `@symbol`                          | `@string.special.symbol`                                                              |
| `@text.*`                          | `@markup.{strong,italic,link,strikethrough}` (with exceptions; see below) |
| `@text.diff.{add,delete,}`         | `@diff.{plus,minus,delta}`                                                            |
| `@text.literal`                    | `@markup.raw`                                                                         |
| `@text.reference`                  | `@markup.link`                                                                        |
| `@text.title`                      | `@markup.heading`                                                                     |
| `@text.uri`                        | `@markup.link.url` (in markup links)                                                  |
| `@text.uri`                        | `@string.special.url` (outside markup)                                                |
| `@text.{todo,warning,note,danger}` | `@comment.{todo,warning,note,error}`                                                  |

</details>

#### Highlight Groups Removed

<details>
<summary>Show groups</summary>
<br />

| Group                              | Comment                                                                               |
| ---------------------------------- | ------------------------------------------------------------------------------------- |
| `VertSplit`                        | use `WinSeparator` instead                                                            |
| `@error`                           | there appears to be no replacement at this time                                       |

</details>

#### Highlight Groups Added

<details>
<summary>Show groups</summary>
<br />

The following is a list of groups which have been added to neovim since the beginning of 2024 and does not include any of the groups mentioned in the previous sections.

| Group                              | Description                                                                           |
| ---------------------------------- | ------------------------------------------------------------------------------------- |
| `@attribute`                       | attribute annotations (e.g. Python decorators, Rust lifetimes)                        |
| `@attribute.builtin`               | builtin annotations (e.g. `@property` in Python)                                      |
| `@comment.documentation`           | comments documenting code                                                             |
| `@function.call`                   | function calls                                                                        |
| `@keyword.conditional.ternary`     | ternary operator (e.g. `?`, `:`)                                                      |
| `@keyword.coroutine`               | keywords related to coroutines (e.g. `go` in Go, `async/await` in Python)             |
| `@keyword.function`                | keywords that define a function (e.g. `func` in Go, `def` in Python)                  |
| `@keyword.modifier`                | keywords defining type modifiers (e.g. `const`, `static`, `public`)                   |
| `@keyword.operator`                | operators that are English words (e.g. `and`, `or`)                                   |
| `@keyword.return`                  | keywords like `return` and `yield`                                                    |
| `@keyword.type`                    | keywords defining composite types (e.g. `struct`, `enum`)                             |
| `@markup.heading.1`                | top-level heading                                                                     |
| `@markup.heading.2`                | section heading                                                                       |
| `@markup.heading.3`                | subsection heading                                                                    |
| `@markup.heading.4`                | and so on                                                                             |
| `@markup.heading.5`                | and so forth                                                                          |
| `@markup.heading.6`                | six levels ought to be enough for anybody                                             |
| `@markup.list.checked`             | checked todo-style list markers                                                       |
| `@markup.list.unchecked`           | unchecked todo-style list markers                                                     |
| `@markup.math`                     | math environments (e.g. `$ ... $` in LaTeX)                                           |
| `@markup.quote`                    | block quotes                                                                          |
| `@markup.raw.block`                | literal or verbatim text as a stand-alone block                                       |
| `@module.builtin`                  | built-in modules or namespaces                                                        |
| `@punctuation.bracket`             | brackets (e.g. `()`, `{}`, `[]`)                                                      |
| `@punctuation.delimiter`           | delimiters (e.g. `;`, `.`, `,`)                                                       |
| `@punctuation.special`             | special symbols (e.g. `{}` in string interpolation)                                   |
| `@string.documentation`            | string documenting code (e.g. Python docstrings)                                      |
| `@string.special.path`             | filenames (e.g. those in .gitignore)                                                  |
| `@tag.attribute`                   | XML-style tag attributes                                                              |
| `@tag.builtin`                     | XML-style tag names (e.g. HTML5 tags)                                                 |
| `@tag.delimiter`                   | XML-style tag delimiters                                                              |
| `@type.builtin`                    | built-in types                                                                        |
| `@variable.builtin`                | built-in variable names (e.g. `this`, `self`)                                         |
| `@variable.parameter.builtin`      | special parameters (e.g. `_`, `it`)                                                   |
| `PmenuMatch`                       | Popup menu: Matched text in normal item.                                              |
| `PmenuSel`                         | Popup menu: Matched text in selected item.                                            |
| `SnippetTabstop`                   | Tabstops in snippets. `:help vim.snippet`                                             |
| `StatusLineTerm`                   | Status line of `terminal` window.                                                     |
| `StatusLineTermNC`                 | Status line of non-current `terminal` windows.                                        |
| `Added`                            | added line in a diff (e.g. `diffAdded` links to this)                                 |
| `Changed`                          | changed line in a diff (e.g. `diffChanged` links to this)                             |
| `Removed`                          | removed line in a diff (e.g. `diffRemoved` links to this)                              |

</details>
